Support remote controlling Ykush devices

edersondisouza opened the pull request:
<!---
Describe what your pull request does,
i.e. fix this bug and how, add a feature, fix documentation…
If you add a feature, please answer these questions:
- what do you use the feature for?
- how does labgrid benefit as a testing library from the feature?
- how did you verify the feature works?
- if hardware is needed for the feature, which hardware is supported and which
  hardware did you test with?
--->
Currently, ykush devices can only be controlled 'locally'. This PR adds support to control them on remote devices. It is based on the seemingly abandoned https://github.com/labgrid-project/labgrid/pull/953 - I basically updated it with the comments there plus a minor fix - first patch, that could go on a different PR if deemed best.

Note that I also removed the 'Export' class created because I'm still figuring out its need - and the commentary for it. If it's deemed necessary, I can bring it back.

<!---
This checklist roughly outlines the steps for new features, remove and add tasks as needed:
--->
**Checklist**
<!---
If you add a driver/resource or modify one:
--->
- [X] The arguments and description in doc/configuration.rst have been updated
- [X] PR has been tested

<!---
In case your PR fixes a bug, please reference it in the next line, i.e.
Fixes #[insert number without brackets here]
--->

jluebbe commented:
> Note that I also removed the 'Export' class created because I'm still figuring out its need - and the commentary for it. If it's deemed necessary, I can bring it back.

The ResourceExport classes are needed to add the host and proxy information which makes local resources accessible via the network. That way, the exporter config can contain a `YKUSHPowerPort` entry, which is then converted to a `NetworkYKUSHPowerPort` by the exporter (which has the correct host and proxy information).

edersondisouza commented:
v2:
- Re-add ResourceExport class for YKUSHPowerPower.

edersondisouza commented:
Any comments about v2, @jluebbe?

jluebbe commented on labgrid/driver/powerdriver.py:
```diff
@@ -271,27 +271,37 @@ def get(self):
-        to control a target's power - https://www.yepkit.com/products/ykush"""
+        to control a target's power - https://www.yepkit.com/products/ykush.
+        Uses ykushcmd tool to control the hub."""
     bindings = {"port": "YKUSHPowerPort", }
```
This needs to explicitly bind to `NetworkYKUSHPowerPort` as well.
```suggestion
    bindings = {"port": {"YKUSHPowerPort", "NetworkYKUSHPowerPort"} }
```
See for example the `FlashromDriver`.

jluebbe commented on labgrid/resource/ykushpowerport.py:
```diff
@@ -15,3 +15,8 @@ class YKUSHPowerPort(Resource):
+
+@target_factory.reg_resource
+@attr.s(eq=False)
+class NetworkYKUSHPowerPort(NetworkResource, YKUSHPowerPort):
```
We don't use multiple inheritance for resources, so die users need explicitly choose which to support.

edersondisouza replied:
Ack.

jluebbe commented on labgrid/remote/client.py:
```diff
@@ -744,6 +746,8 @@ def power(self):
                     drv = self._get_driver_or_new(target, "PDUDaemonDriver", name=name)
                 elif isinstance(resource, TasmotaPowerPort):
                     drv = self._get_driver_or_new(target, "TasmotaPowerDriver", name=name)
+                elif isinstance(resource, YKUSHPowerPort):
```
```suggestion
                elif isinstance(resource, NetworkYKUSHPowerPort):
```

jluebbe commented on labgrid/remote/client.py:
```diff
@@ -727,13 +727,15 @@ def power(self):
         from ..resource.power import NetworkPowerPort, PDUDaemonPort
         from ..resource.remote import NetworkUSBPowerPort, NetworkSiSPMPowerPort
-        from ..resource import TasmotaPowerPort
+        from ..resource import TasmotaPowerPort, YKUSHPowerPort
```
```suggestion
        from ..resource import TasmotaPowerPort, NetworkYKUSHPowerPort
```
Also add the import the the resource module.

jluebbe reviewed (changes requested):
Please also add a hint to the changelog that it's now necessary to install the ykushcmd tool.

In `doc/configuration.rst`, there's still a recommendation to use `pykush -l` for the serial. Is there an alternative with ykushcmd?

edersondisouza commented:
v3:
  - No more inheritance from `YKUSHPowerPort` in `NetworkYKUSHPowerPort`;
  - Using `NetworkYKUSHPowerPort` directly on `client.py`;
  - `YKUSHPowerDriver` explicitly binds to `NetworkYKUSHPowerPort`;
  - Updated `configuration.rst` example to use `ykushcmd`;
  - Updated commit message to inform about installing `ykushcmd`.

edersondisouza commented:
Any more coments, @jluebbe?

edersondisouza commented:
Gentle reminder about this PR, @jluebbe =D
